src/rulesets/Base/mapreduce.jl

```diff
@@ -153,6 +153,32 @@ for Config in (RuleConfig, RuleConfig{>:HasReverseMode})
     end
 end
 
+#####
+##### `cumsum`
+#####
+
+function frule((_, xdot), ::typeof(cumsum), x::AbstractArray; dims::Integer)
```

Comment:
Perhaps an `frule` for `cumsum!` here as well?

Reply:
Are forward rlues for mutating functions safe? I haven't thought that through.

Reply:
For reverse mode, https://github.com/JuliaDiff/ChainRules.jl/pull/521 has thorny examples.

Reply:
Yes, I believe so, since the primal and tangent are mutated simultaneously. We already have several `frule`s for mutating functions. See e.g. https://github.com/JuliaDiff/ChainRules.jl/blob/main/src/rulesets/LinearAlgebra/factorization.jl.

Reply:
Ok, nice.

Then someone should do a big PR adding this to half the Base functions, right? `sum!` and `circshift!` et. al.

Reply:
Must we define methods like `cumsum!([0,0,0], ZeroTangent())`?

Reply:
Hm, not certain. @oxinabox didn't we decide at some point that it's up to the AD to handle cases where all (co)tangents are zero tangents?

Reply:
I guess we can solve this when or if it comes up.

Reply:
yeah we expect the AD to handle cases in reverse mode when inputs are all `ZeroTangent`.
For forwards mode, i would need to think more, but i think even with mutation https://en.wikipedia.org/wiki/Nothing_comes_from_nothing

We can absolutely add fixes as things come up.